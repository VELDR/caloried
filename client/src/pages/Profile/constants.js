export const EDIT_PROFILE = 'Profile/EDIT_PROFILE';
