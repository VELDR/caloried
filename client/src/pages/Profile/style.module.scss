@import '@styles/base';

.page {
  margin-bottom: 7rem;
  @include apply-at-sm-size {
    margin: 0;
  }
  .container {
    .title {
      text-align: center;
      width: 100%;
      border-bottom: 1px solid var(--color-text-secondary);
      font-weight: 700;
      font-size: $font-size-larger;
      padding: 1rem 0.5rem;
      @include apply-at-sm-size {
        text-align: left;
      }
    }
    .header {
      padding: 1rem 0;
      .avatar {
        position: relative;
        margin: 0 auto;
        width: fit-content;
        padding: 1rem;
        &__image {
          @include box(7rem);
        }
        &__edit {
          cursor: pointer;
          position: absolute;
          bottom: 15px;
          right: 20px;
          padding: 0.5rem;
          border-radius: 50%;
          background-color: var(--color-accent);
          &:hover {
            background-color: var(--color-text-primary);
            color: var(--color-bg-primary);
          }
        }
      }
      .heading {
        text-align: center;
        &__username {
          font-size: $font-size-large;
          font-weight: 700;
        }
        &__email {
          color: var(--color-text-secondary);
        }
        &__goal {
          padding: 0.5rem;
          @include flex-center;
          text-transform: capitalize;
        }
        .gainColor {
          color: var(--color-primary);
        }
        .maintainColor {
          color: $color-fat;
        }
        .loseColor {
          color: $color-error;
        }
      }
    }

    .body {
      border-top: 1px solid var(--color-text-secondary);
      .section {
        border-bottom: 1px solid var(--color-text-secondary);
        text-transform: uppercase;
        font-weight: 700;
        padding: 0.25rem 0.5rem;
      }
      .profile {
        border-bottom: 1px solid var(--color-text-secondary);
        padding: 0.5rem;
        &__label {
          text-transform: uppercase;
          font-weight: 700;
          font-size: $font-size-small;
          color: var(--color-text-secondary);
        }
        &__info {
          span {
            text-transform: capitalize;
          }
        }
      }
      .nutrition {
        .nutrient {
          border-bottom: 1px solid var(--color-text-secondary);
          padding: 0.5rem;
          &__label {
            text-transform: uppercase;
            font-weight: 700;
            font-size: $font-size-small;
            color: var(--color-text-secondary);
          }
          &__info {
            @include flex-items-center(0.25rem);
            font-weight: 700;
            span {
              font-size: $font-size-small;
              font-style: italic;
            }
          }
        }

        .icon {
          @include box(1.25rem);
        }
      }
    }
  }
}
