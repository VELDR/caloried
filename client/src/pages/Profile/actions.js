import { CHANGE_PASSWORD, EDIT_PROFILE } from './constants';

export const editProfile = (data, token, callback) => ({
  type: EDIT_PROFILE,
  data,
  token,
  callback,
});

export const changePassword = (data, token, callback) => ({
  type: CHANGE_PASSWORD,
  data,
  token,
  callback,
});
