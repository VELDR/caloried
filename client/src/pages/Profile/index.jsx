import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Avatar } from '@mui/material';
import { Balance, Edit, KeyboardDoubleArrowDown, KeyboardDoubleArrowUp } from '@mui/icons-material';
import proteinIcon from '@static/images/protein.png';
import fatIcon from '@static/images/fat.png';
import carbsIcon from '@static/images/carbs.png';
import caloriesIcon from '@static/images/calories.svg';
import { getActivityLevel } from '@utils/formatUtils';
import { calculateAge } from '@utils/calculateUtils';

import EditProfileModal from '@components/profile/EditProfileModal';
import { getUser } from '@pages/Diary/actions';
import { selectUser } from '@pages/Diary/selectors';
import { selectToken } from '@containers/Client/selectors';

import classes from './style.module.scss';

const Profile = ({ user, token }) => {
  const dispatch = useDispatch();

  const [isEditOpen, setIsEditOpen] = useState(false);

  useEffect(() => {
    if (token) {
      dispatch(getUser(token));
    }
  }, [dispatch, token]);

  const handleEditClick = () => {
    setIsEditOpen(true);
  };

  const handleCloseEditModal = () => {
    setIsEditOpen(false);
  };

  const getGoalInfo = (goal) => {
    switch (goal) {
      case 'gain':
        return { icon: <KeyboardDoubleArrowUp />, text: 'Gaining Weight', className: classes.gainColor };
      case 'maintain':
        return { icon: <Balance />, text: 'Maintaining Weight', className: classes.maintainColor };
      case 'lose':
        return { icon: <KeyboardDoubleArrowDown />, text: 'Losing Weight', className: classes.loseColor };
      default:
        return { icon: null, text: '', className: '' };
    }
  };

  const goalContent = getGoalInfo(user?.goal);

  return (
    <div className={classes.page}>
      <div className={classes.container}>
        <div className={classes.title}>My Profile</div>
        <div className={classes.header}>
          <div className={classes.avatar}>
            <Avatar src={`${import.meta.env.VITE_API_BASE_URL}${user?.avatar}`} className={classes.avatar__image} />
            <Edit className={classes.avatar__edit} onClick={handleEditClick} />
          </div>
          <div className={classes.heading}>
            <div className={classes.heading__username}>{user?.username}</div>
            <div className={classes.heading__email}>{user?.email}</div>
            <div className={`${classes.heading__goal} ${goalContent.className}`}>
              {goalContent.icon} {goalContent.text}
            </div>
          </div>
        </div>

        <div className={classes.body}>
          <div className={classes.section}>Physical Profile</div>
          <div className={classes.profile}>
            <div className={classes.profile__label}>Sex</div>
            <div className={classes.profile__info}>
              <span>{user?.sex}</span>
            </div>
          </div>
          <div className={classes.profile}>
            <div className={classes.profile__label}>Age</div>
            <div className={classes.profile__info}>{calculateAge(user?.dob)} years old </div>
          </div>
          <div className={classes.profile}>
            <div className={classes.profile__label}>Height</div>
            <div className={classes.profile__info}>{user?.height} cm</div>
          </div>
          <div className={classes.profile}>
            <div className={classes.profile__label}>Weight</div>
            <div className={classes.profile__info}>{user?.weight} kg</div>
          </div>
          <div className={classes.profile}>
            <div className={classes.profile__label}>Activity Level</div>
            <div className={classes.profile__info}>{getActivityLevel(user?.activityLevel)}</div>
          </div>
          <div className={classes.section}>Nutritional Intake</div>
          <div className={classes.nutrition}>
            <div className={classes.nutrient}>
              <div className={classes.nutrient__label}>Basal Metabolic Rate (BMR)</div>
              <div className={classes.nutrient__info}>
                <img src={caloriesIcon} alt="" className={classes.icon} />
                {user?.bmr} <span>kcal/day</span>
              </div>
            </div>
            <div className={classes.nutrient}>
              <div className={classes.nutrient__label}>Protein Intake</div>
              <div className={classes.nutrient__info}>
                <img src={proteinIcon} alt="" className={classes.icon} />
                {user?.proteinIntake} <span>g/day</span>
              </div>
            </div>
            <div className={classes.nutrient}>
              <div className={classes.nutrient__label}>Carbohydrate Intake</div>
              <div className={classes.nutrient__info}>
                <img src={carbsIcon} alt="" className={classes.icon} />
                {user?.carbsIntake}
                <span>g/day</span>
              </div>
            </div>
            <div className={classes.nutrient}>
              <div className={classes.nutrient__label}>Fat Intake</div>
              <div className={classes.nutrient__info}>
                <img src={fatIcon} alt="" className={classes.icon} />
                {user?.fatIntake}
                <span>g/day</span>
              </div>
            </div>
          </div>
        </div>
      </div>
      {user && <EditProfileModal open={isEditOpen} onClose={handleCloseEditModal} user={user} token={token} />}
    </div>
  );
};

Profile.propTypes = {
  user: PropTypes.object,
  token: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  user: selectUser,
  token: selectToken,
});

export default connect(mapStateToProps)(Profile);
