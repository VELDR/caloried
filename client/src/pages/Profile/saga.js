import { changePasswordApi, editProfileApi } from '@domain/api';
import toast from 'react-hot-toast';
import { call, put, takeLatest } from 'redux-saga/effects';
import { setLoading, showPopup } from '@containers/App/actions';
import { setUser } from '@pages/Diary/actions';
import { CHANGE_PASSWORD, EDIT_PROFILE } from './constants';

export function* doEditProfile({ data, token, callback }) {
  yield put(setLoading(true));
  try {
    const response = yield call(editProfileApi, data, token);
    yield put(setUser(response.user));
    yield call(callback);
    toast.success(response.message);
  } catch (error) {
    if (error.response && error.response.data) {
      toast.error(error.response.data.message);
    } else {
      yield put(showPopup());
    }
  } finally {
    yield put(setLoading(false));
  }
}

export function* doChangePassword({ data, token, callback }) {
  yield put(setLoading(true));
  try {
    const response = yield call(changePasswordApi, data, token);
    yield call(callback);
    toast.success(response.message);
  } catch (error) {
    toast.error(error.response.data.message);
  } finally {
    yield put(setLoading(false));
  }
}

export default function* profileSaga() {
  yield takeLatest(EDIT_PROFILE, doEditProfile);
  yield takeLatest(CHANGE_PASSWORD, doChangePassword);
}
