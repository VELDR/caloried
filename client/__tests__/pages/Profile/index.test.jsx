import Profile from '@pages/Profile';
import { render, fireEvent } from '@utils/testHelper';

jest.mock('@static/images/protein.png', () => 'path/to/protein.png');
jest.mock('@static/images/carbs.png', () => 'path/to/carbs.png');
jest.mock('@static/images/fat.png', () => 'path/to/fat.png');

let wrapper;

beforeEach(() => {
  wrapper = render(<Profile />);
});

describe('Profile Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId } = wrapper;

    expect(getByTestId('profile')).toBeInTheDocument();
  });

  it('should open the change password modal when the Change Password button is clicked', () => {
    const { getByTestId } = wrapper;

    fireEvent.click(getByTestId('change-password-button'));

    expect(getByTestId('change-password-modal')).toBeInTheDocument();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
