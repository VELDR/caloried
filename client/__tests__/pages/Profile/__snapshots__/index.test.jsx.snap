// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div>
      <div
        data-testid="profile"
      >
        <div>
          <div>
            Profil Saya
          </div>
          <div>
            <div>
              <div
                class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
              >
                <img
                  class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="http://localhost:8080undefined"
                />
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="edit-profile-button"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </div>
            <div>
              <div />
              <div />
              <div
                class="undefined "
              >
                 
              </div>
            </div>
          </div>
          <div>
            <div>
              Profil Fisik
            </div>
            <div>
              <div>
                Jenis Kelamin
              </div>
              <div>
                <span />
              </div>
            </div>
            <div>
              <div>
                Usia
              </div>
              <div>
                NaN
                 
                tahun
              </div>
            </div>
            <div>
              <div>
                Tinggi Badan
              </div>
              <div>
                 cm
              </div>
            </div>
            <div>
              <div>
                Berat Badan
              </div>
              <div>
                 kg
              </div>
            </div>
            <div>
              <div>
                Tingkat Aktivitas
              </div>
              <div>
                Unknown Activity Level
              </div>
            </div>
            <div>
              Asupan Nutrisi
            </div>
            <div>
              <div>
                <div>
                  Asupan Kalori
                </div>
                <div>
                  <img
                    alt=""
                  />
                   
                  <span>
                    kcal/
                    hari
                  </span>
                </div>
              </div>
              <div>
                <div>
                  Asupan Protein
                </div>
                <div>
                  <img
                    alt=""
                    src="path/to/fat.png"
                  />
                   
                  <span>
                    g/
                    hari
                  </span>
                </div>
              </div>
              <div>
                <div>
                  Asupan Karbohidrat
                </div>
                <div>
                  <img
                    alt=""
                    src="path/to/fat.png"
                  />
                  <span>
                    g/
                    hari
                  </span>
                </div>
              </div>
              <div>
                <div>
                  Asupan Lemak
                </div>
                <div>
                  <img
                    alt=""
                    src="path/to/fat.png"
                  />
                  <span>
                    g/
                    hari
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              data-testid="change-password-button"
            >
              Ubah Sandi
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="profile"
    >
      <div>
        <div>
          Profil Saya
        </div>
        <div>
          <div>
            <div
              class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
            >
              <img
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="http://localhost:8080undefined"
              />
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="edit-profile-button"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </div>
          <div>
            <div />
            <div />
            <div
              class="undefined "
            >
               
            </div>
          </div>
        </div>
        <div>
          <div>
            Profil Fisik
          </div>
          <div>
            <div>
              Jenis Kelamin
            </div>
            <div>
              <span />
            </div>
          </div>
          <div>
            <div>
              Usia
            </div>
            <div>
              NaN
               
              tahun
            </div>
          </div>
          <div>
            <div>
              Tinggi Badan
            </div>
            <div>
               cm
            </div>
          </div>
          <div>
            <div>
              Berat Badan
            </div>
            <div>
               kg
            </div>
          </div>
          <div>
            <div>
              Tingkat Aktivitas
            </div>
            <div>
              Unknown Activity Level
            </div>
          </div>
          <div>
            Asupan Nutrisi
          </div>
          <div>
            <div>
              <div>
                Asupan Kalori
              </div>
              <div>
                <img
                  alt=""
                />
                 
                <span>
                  kcal/
                  hari
                </span>
              </div>
            </div>
            <div>
              <div>
                Asupan Protein
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                 
                <span>
                  g/
                  hari
                </span>
              </div>
            </div>
            <div>
              <div>
                Asupan Karbohidrat
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                <span>
                  g/
                  hari
                </span>
              </div>
            </div>
            <div>
              <div>
                Asupan Lemak
              </div>
              <div>
                <img
                  alt=""
                  src="path/to/fat.png"
                />
                <span>
                  g/
                  hari
                </span>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            data-testid="change-password-button"
          >
            Ubah Sandi
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
